RFS/src/main/java/org/opensearch/migrations/bulkload/lucene/LuceneDocumentsReader9.java

```diff
@@ -0,0 +1,230 @@
+     * loses synchronization, and re-joins.  They make it more likely the cluster can just replay the missed operations.  You
+     * can read a bit more about soft deletes here:
+     *
+     * https://www.elastic.co/guide/en/elasticsearch/reference/7.10/index-modules-history-retention.html
+     *
+     * Soft deletes works by having the application writing the Lucene Index define a field that is used to mark a document as
+     * "soft deleted" or not.  When a document is marked as "soft deleted", it is not returned in search results, but it is
+     * still present in the Lucene Index.  The status of whether any given document is "soft deleted" or not is stored in the
+     * Lucene Index itself.  By default, Elasticsearch 7+ Indices have soft deletes enabled;  this is an Index-level setting.
+     * Just like deleted documents and old versions of updated documents, we don't want to reindex them agaisnt the target
+     * cluster.
+     *
+     * In order to retrieve only those documents that would be considered "live" in ES/OS, we use a few tricks:
+     * 1. We make sure we use the latest Lucene commit point on the Lucene Index.  A commit is a Lucene abstraction that
+     *     comprises a consistent, point-in-time view of the Segments in the Lucene Index.  By default, a DirectoryReader
+     *     will use the latest commit point.
+     * 2. We use a concept called "liveDocs" to determine if a document is "live" or not.  The liveDocs are a bitset that
+     *    is associated with each Lucene segment that tells us if a document is "live" or not.  If a document is hard-deleted
+     *    (i.e. deleted from the ES/OS perspective or an old version of an updated doc), then the liveDocs bit for that
+     *    document will be false.
+     * 3. We wrap our DirectoryReader in a SoftDeletesDirectoryReaderWrapper if it's possible that soft deletes will be
+     *    present in the Lucene Index.  This wrapper will filter out documents that are marked as "soft deleted" in the
+     *    Lucene Index.
+
+     */
+    public Flux<RfsLuceneDocument> readDocuments(int startSegmentIndex, int startDoc) {
+        return Flux.using(
+            () -> wrapReader(getReader(), softDeletesPossible, softDeletesField),
+            reader -> readDocsByLeavesFromStartingPosition(reader, startSegmentIndex, startDoc),
+            reader -> {
+                try {
+                    reader.close();
+                } catch (IOException e) {
+                    log.atError().setMessage("Failed to close DirectoryReader").setCause(e).log();
+                    throw Lombok.sneakyThrow(e);
+                }
+        });
+    }
+
+    protected DirectoryReader getReader() throws IOException {
+        try (var directory = FSDirectory.open(indexDirectoryPath)) {
+            var commits = DirectoryReader.listCommits(directory);
+            var latestCommit = commits.get(commits.size() - 1);
+
+            return DirectoryReader.open(latestCommit, 0, null);
```

Comment:
Do we want to set the minimum Lucene version here to Lucene 8, given we're explicitly fullfilling Lucene 7 and before with separate readers?  Feels like a good safety mechanism to ensure we aren't "surprised" by the wrong LuceneDocumentsReader class being used.